File: rust/examples/pagination.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// pagination: Demonstrate fetch_all and Paginator for paginated API results.
//
// Usage: pagination [host_url]

use ghidra::paginate::{fetch_all, Paginator};
use libghidra as ghidra;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. fetch_all: collect every function in one call
    println!("=== fetch_all: all functions ===\n");
    let all_funcs = fetch_all(|limit, offset| {
        let resp = client.list_functions(0, u64::MAX, limit, offset)?;
        Ok(resp.functions)
    })
    .unwrap_or_else(|e| {
        eprintln!("fetch_all(list_functions) failed: {e}");
        std::process::exit(1);
    });
    println!("Total functions: {}", all_funcs.len());
    for (i, f) in all_funcs.iter().enumerate().take(10) {
        println!(
            "  [{:>3}] 0x{:x}  {}  ({} bytes)",
            i, f.entry_address, f.name, f.size
        );
    }
    if all_funcs.len() > 10 {
        println!("  ... and {} more", all_funcs.len() - 10);
    }

    // 2. Paginator with custom page size: iterate symbols page by page
    println!("\n=== Paginator: symbols (page_size=25) ===\n");
    let paginator = Paginator::new(|limit, offset| {
        let resp = client.list_symbols(0, u64::MAX, limit, offset)?;
        Ok(resp.symbols)
    })
    .page_size(25);

    let mut page_num = 0;
    let mut total_symbols = 0;
    for page in paginator {
        let items = page.unwrap_or_else(|e| {
            eprintln!("Paginator page failed: {e}");
            std::process::exit(1);
        });
        if items.is_empty() {
            break;
        }
        page_num += 1;
        total_symbols += items.len();
        println!(
            "Page {}: {} symbols (first: '{}', last: '{}')",
            page_num,
            items.len(),
            items.first().map_or("?", |s| &s.name),
            items.last().map_or("?", |s| &s.name),
        );
        // Stop after 5 pages for demonstration purposes
        if page_num >= 5 {
            println!("  (stopping after 5 pages for demo)");
            break;
        }
    }
    println!(
        "\nSymbols seen across {} pages: {}",
        page_num, total_symbols
    );

    // 3. fetch_all for signatures: collect all function signatures
    println!("\n=== fetch_all: all function signatures ===\n");
    let all_sigs = fetch_all(|limit, offset| {
        let resp = client.list_function_signatures(0, u64::MAX, limit, offset)?;
        Ok(resp.signatures)
    })
    .unwrap_or_else(|e| {
        eprintln!("fetch_all(list_function_signatures) failed: {e}");
        std::process::exit(1);
    });
    println!("Total signatures: {}", all_sigs.len());
    for sig in all_sigs.iter().take(5) {
        println!(
            "  0x{:x}  {} -> {}",
            sig.function_entry_address, sig.function_name, sig.prototype
        );
    }
    if all_sigs.len() > 5 {
        println!("  ... and {} more", all_sigs.len() - 5);
    }

    // 4. Compare counts
    println!("\n=== Summary ===");
    println!("  Functions:  {}", all_funcs.len());
    println!("  Signatures: {}", all_sigs.len());
    println!(
        "  Symbols:    {}+ (first {} pages)",
        total_symbols, page_num
    );
}

File: python/examples/pagination.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
"""pagination: demonstrate fetching every page of a paginated list RPC.

Mirrors rust/examples/pagination.rs. The Python client doesn't ship a
dedicated Paginator class — the simple loop below is the idiomatic
equivalent.

Usage: python pagination.py [host_url]
"""

from __future__ import annotations

import sys

import libghidra as ghidra


def fetch_all(call, page_size: int = 100):
    """Collect every page from a paginated list RPC.

    `call(limit, offset)` should return the items vector for one page.
    """
    items: list = []
    offset = 0
    while True:
        page = call(page_size, offset)
        if not page:
            break
        items.extend(page)
        offset += len(page)
        if len(page) < page_size:
            break
    return items


def main() -> int:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    print("=== fetch_all: all functions ===\n")
    all_funcs = fetch_all(
        lambda limit, offset: client.list_functions(
            range_start=0, range_end=2**64 - 1, limit=limit, offset=offset
        ).functions
    )
    print(f"Total functions: {len(all_funcs)}")
    for i, f in enumerate(all_funcs[:10]):
        print(f"  [{i:>3}] 0x{f.entry_address:x}  {f.name}  ({f.size} bytes)")
    if len(all_funcs) > 10:
        print(f"  ... and {len(all_funcs) - 10} more")

    print("\n=== page-by-page: symbols (page_size=25) ===\n")
    page_num = 0
    total_symbols = 0
    offset = 0
    while page_num < 5:
        items = client.list_symbols(
            range_start=0, range_end=2**64 - 1, limit=25, offset=offset
        ).symbols
        if not items:
            break
        page_num += 1
        total_symbols += len(items)
        first = items[0].name if items else "?"
        last = items[-1].name if items else "?"
        print(
            f"Page {page_num}: {len(items)} symbols "
            f"(first: '{first}', last: '{last}')"
        )
        offset += len(items)

    print("\n=== fetch_all: all function signatures ===\n")
    all_sigs = fetch_all(
        lambda limit, offset: client.list_function_signatures(
            range_start=0, range_end=2**64 - 1, limit=limit, offset=offset
        ).signatures
    )
    print(f"Total signatures: {len(all_sigs)}")
    for sig in all_sigs[:5]:
        print(
            f"  0x{sig.function_entry_address:x}  "
            f"{sig.function_name} -> {sig.prototype}"
        )
    if len(all_sigs) > 5:
        print(f"  ... and {len(all_sigs) - 5} more")

    print("\n=== Summary ===")
    print(f"  Functions:  {len(all_funcs)}")
    print(f"  Signatures: {len(all_sigs)}")
    print(f"  Symbols:    {total_symbols}+ (first {page_num} pages)")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
